### PR TITLE
HAL_Linux: added SRXL decoder and improved DSM/ST24/SUMD decoders

### PR DESCRIPTION
**Pull request by tridge**
this adds a SRXL decoder for newer Spektrum receivers (such as AR7700). 
It also adds logic to "lock onto" a RC input type in the RCInput_115200 decoder

### CODE REVIEW

File: libraries/AP_HAL/utility/srxl.cpp [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+            // special handling for channel 12
+            // see http://www.deviationtx.com/forum/protocol-development/2088-18-channels-for-dsm2-dsmx?start=40
+            //printf("c12: 0x%x %02x %02x\n", (unsigned)(b>>9), (unsigned)buffer[0], (unsigned)buffer[1]);
+            v = (b & 0x1FF) * 4;
+            c = 12 + (((b >> 9) & 0x3) | ((buffer[1]&1)<<2));
+        }
+
+        // if channel number if greater than 16 then it is a X-Plus
+        // channel. We don't yet know how to decode those. There is some information here:
+        // but we really need some sample data to confirm
```
Comment (OXINARF) [inline libraries/AP_HAL/utility/srxl.cpp]: Left overs? It can decode up to 20 channels right?

Comment (tridge) [inline libraries/AP_HAL/utility/srxl.cpp]: I've now fixed the XPlus channel decoding (thanks to load of a DX18 transmitter today)

**Comment by tridge** [comment]
we can make it an example prog, but I'll do that as a separate PR when we go to a table driver approach as we discussed on gitter

**Comment by tridge** [comment]
applied

File: libraries/AP_HAL/utility/srxl.cpp (hunk 1 above)
Comment (OXINARF) [inline libraries/AP_HAL/utility/srxl.cpp]: I think you haven't read the comment in code. It says we can't decode more than 16 channels.

Comment (tridge) [inline libraries/AP_HAL/utility/srxl.cpp]: sorry Francisco, fixed now

**Pull request closed without merging**